packages/core/src/rules/common/assertions/utils.ts

```diff
@@ -0,0 +1,148 @@
+  suggest?: string[];
+  runsOnKeys: boolean;
+  runsOnValues: boolean;
+};
+
+export function buildVisitorObject(subject: string, context: Record<string, any>[], subjectVisitor: any) {
+  if (!context) {
+    return  { [subject]: subjectVisitor };
+  }
+
+  let currentVisitorLevel: Record<string, any> = {};
+  const visitor: Record<string, any> = currentVisitorLevel;
+
+  for (let index=0; index < context.length; index++) {
+    const node = context[index];
+    if (context.length === index + 1 && node.type === subject) {
+      // no need to create separate visitor for the last element
+      // which is the same as subject;
+      // will check includes/excludes it in the last visitor.
+      continue;
+    }
+    const matchParentKeys = node.matchParentKeys;
+    const excludeParentKeys = node.excludeParentKeys;
+
+    if (matchParentKeys && excludeParentKeys) {
+      throw new Error(`Both 'matchParentKeys' and 'excludeParentKeys' can't be under one context item`);
+    }
+
+    if (matchParentKeys || excludeParentKeys) {
+      currentVisitorLevel[node.type] = {
+        skip: (_value: any, key: string) => {
+          if (matchParentKeys) {
+            return !matchParentKeys.includes(key);
+          }
+          if (excludeParentKeys) {
+            return excludeParentKeys.includes(key);
+          }
+        }
+      }
+    } else {
+      currentVisitorLevel[node.type] = {}
+    }
+    currentVisitorLevel = currentVisitorLevel[node.type];
+  }
+
+  currentVisitorLevel[subject] = subjectVisitor;
+
+  return visitor;
+}
+
+export function buildSubjectVisitor(properties: string | string[], asserts: Assert[], context?: Record<string, any>[]) {
+  return function(node: any, { report, location, key, type }: UserContext) {
+
+    // check context for same node type parent includes/excludes:
+    if (context) {
+      const lastContextNode = context[context.length-1];
+      if (lastContextNode.type === type.name) {
+        const matchParentKeys = lastContextNode.matchParentKeys;
+        const excludeParentKeys = lastContextNode.excludeParentKeys;
+
+        if (matchParentKeys && !matchParentKeys.includes(key)) {
+          return;
+        }
+        if (excludeParentKeys && excludeParentKeys.includes(key)) {
+          return;
+        }
+      }
+    }
+
+    if (properties) {
+      properties = Array.isArray(properties) ? properties : [properties];
+    }
+
+    for (const assert of asserts) {
+      if (properties) {
+        for (const property of properties) {
+          runAssertion(node[property], assert, location.child(property), report);
+        }
+      } else {
+        runAssertion(Object.keys(node), assert, location.key(), report);
+      }
+    }
+  }
+}
+
+export const getIntersectionLength = (keys: string[], properties: string[]): number => {
+  const props = new Set(properties);
+  let count = 0;
+  for (const key of keys) {
+    if (props.has(key)) {
+      count++;
+    }
+  }
+  return count;
```

Comment:
I would keep the original due to perf reasons.
This function is on the hot-path so performance matters.

Reply:
The performance should be about the same, 6-7% difference according to https://jsbench.me. Is this really worth it?

Reply:
Yes. This is on hot path, this function may be executed for every value in the api definition (multiple times). So even a small difference matters here.

Especially when readability is still not too bad.

We can add extra benchmark for assertions later and see if it really matters (it may not in fact). But i would keep it as is for now.